### PR TITLE
Refactoring refinerycms-search

### PR DESCRIPTION
**Pull request by bricesanchez**
- [x] Update README
- [x] Add scope to exclude draft records
- [x] Improve i18n of content
- [x] Add pagination support
- [x] Add configuration initializer
- [x] Ability to use custom routes
- [x] Namespaced the extension

It's ready to be merged after your review and a little squash of commits

### CODE REVIEW

**Comment by bricesanchez** [comment]
I need help to find why i've this error : https://travis-ci.org/refinery/refinerycms-search/jobs/57894663

```bash
/home/travis/.rvm/gems/ruby-2.2.0/bundler/gems/refinerycms-04cc80fc61bb/core/lib/refinery.rb:189:in `validate_extension!': Engine must define a root accessor that returns a pathname to its root (Refinery::InvalidEngineError)
	from /home/travis/.rvm/gems/ruby-2.2.0/bundler/gems/refinerycms-04cc80fc61bb/core/lib/refinery.rb:51:in `register_extension'
```

**Comment by anitagraham** [comment]
I have come across this problem (root not being defined) again today.
It seems that `engine.rb` is calling `register_extension` before the code in which `root` is defined is called.

I can make it work in my extension by moving the lines around thus:

````ruby
module Components

      class << self
        def root
          @root ||= Pathname.new(File.expand_path('../../../', __FILE__))
        end

        def factory_paths
          @factory_paths ||= [ root.join('spec', 'factories').to_s ]
        end

        require 'refinery/caststone/components/engine'
        require 'refinery/caststone/components/configuration'
      end
    end
````
I need to look at the order of these initialization steps to work out what is really happening in this case.

**Comment by anitagraham** [comment]
Followup: interesting but not helpful.
I fixed some other deeper problems. Everything works.
Moved the `require` statements etc back to the way they were.
Everything still works.

So this error has been (a couple of times for me) a sign of things going wrong elsewhere in the extension.

**Comment by bricesanchez** [comment]
Thanks @anitagraham ! It helped me to find that my config accessor enable_for was bad. I'm working on the PR.

**Comment by bricesanchez** [comment]
Is someone could review this PR ? Thanks !

**Comment by bricesanchez** [comment]
Sorry something went wrong with my previous commits, all should be ok now to be merge.

**Comment by bricesanchez** [comment]
@parndt do you have time to review this PR? :)

**Comment by parndt** [comment]
37 commits?!

File: app/controllers/refinery/search/search_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+module Refinery
+  module Search
+    class SearchController < ::ApplicationController
+
+      before_filter :find_page
+
+      def show
+        @results = Refinery::Search::SearchEngine.search(params[:query])
+        @results = @results.paginate(page: params[:page], per_page: Refinery::Search.results_per_page)
+      end
```
Comment (parndt) [inline app/controllers/refinery/search/search_controller.rb]: This could be a `before_action`, but it is also totally fine as it is.

File: app/views/refinery/search/search/show.html.erb [hunk 2]
```diff
@@ -1,10 +1,16 @@
 <% content_for :body_content_title do %>
-  <%= t('.search_results_for', :what => params[:query]) %>
+  <% if params[:query] && params[:query].empty? %>
+    <%= t('.no_results') %>
+  <% else %>
+    <%= t('.search_results_for', what: params[:query]) %>
+  <% end %>
```
Comment (parndt) [inline app/views/refinery/search/search/show.html.erb]: I'd prefer this were positive:

```erb
<% if params[:query].present? %>
  <%= t('.search_results_for', what: params[:query]) %>
<% else %>
  <%= t('.no_results') %>
<% end %>
```

Same with the body block

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,5 @@
 Refinery::Core::Engine.routes.draw do
-  get "/search", :to => 'search#show', :as => 'search'
-end
+  namespace :search, path: Refinery::Search.page_url do
+    get '/' , to: 'search#show'
```
Comment (parndt) [inline config/routes.rb]: Does this produce the same as using `root`?

```ruby
namespace :search, path: Refinery::Search.page_url do
  root to: 'search#show'
end
```

?

File: readme.md [hunk 4]
```diff
@@ -1,14 +1,34 @@
-# Search plugin for [Refinery CMS](http://www.refinerycms.com)
+# Refinery CMS Search [![Build Status](https://travis-ci.org/refinery/refinerycms-search.svg?branch=master)](https://travis-ci.org/refinery/refinerycms-search)
+
+Simple search engine for [Refinery CMS](http://refinerycms.com). By default, it supports search in `Refinery::Page` and `Refinery::Blog::Post` (if installed), it can be easily configured for another Rails models.
```
Comment (parndt) [inline readme.md]: Grammar:

```
Simple search engine for [Refinery CMS](http://refinerycms.com). By default it supports search in `Refinery::Page`, and `Refinery::Blog::Post` (if installed), but it can be easily configured for other Rails models too.
```

File: readme.md [hunk 5]
```diff
@@ -1,14 +1,34 @@
...
+
+This version of `refinerycms-search` supports Refinery 3.x and Rails 4.1.x.  To use Rails 3.x / Refinery 2.1.x use the [refinerycms-search "Refinery CMS 2-1 stable branch"](http://github.com/refinery/refinerycms-search/tree/2-1-stable).
```
Comment (parndt) [inline readme.md]: It'd be nice to order these in the same way.

```
This version of `refinerycms-search` supports Refinery 3.x and Rails 4.1.x.  
For Refinery 2.1.x and Rails 3.x, use the [2-1-stable branch](http://github.com/refinery/refinerycms-search/tree/2-1-stable).
```

File: readme.md [hunk 6]
```diff
@@ -1,14 +1,34 @@
...
 Powered by: [acts_as_indexed](https://github.com/dougal/acts_as_indexed) and [refinerycms-acts-as-indexed](https://github.com/refinery/refinerycms-acts-as-indexed) -
 Check his readme and documentation for more info on how it works.
 
+
+### In summary you can
+
+* Search in `Refinery::Page` and `Refinery::Blog::Post` models ;
+* Add others models in your configuration ;
+* Scope search query with a special method ;
+* Use a custom search page url ;
```
Comment (parndt) [inline readme.md]: URL is an acronym so should be `URL`, not `url`.
I'd personally remove the semicolons too: as it's already a list.

File: readme.md [hunk 7]
```diff
@@ -1,14 +1,34 @@
...
+* Paginate the search result.
+
+What it can't do :
+* It can't search all languages in the same query. it only display the results of the current locale.
```
Comment (parndt) [inline readme.md]: The fullstop should have a capital letter after it, or it should not be a fullstop.

I'd use a colon:

```
* It can't search all languages in the same query: it only display results for the current locale.
```

File: readme.md [hunk 8]
```diff
@@ -25,28 +45,35 @@ Finally seed your database and you're done.
...
 
-## RE-SAVE all records that have not been indexed before.
+## Index records
+
+You will have to **RE-SAVE** all records that have not been indexed before.
+
+You can do it easily with a command like this in rails console :
+
+    Refinery::Page.all(&:save)
+
```
Comment (parndt) [inline readme.md]: This code block could be syntax highlighted:

<pre>
```ruby
Refinery::Page.all(&:save)
```
</pre>

File: refinerycms-search.gemspec [hunk 9]
```diff
@@ -8,6 +8,7 @@ Gem::Specification.new do |s|
   s.homepage          = %q{http://refinerycms.com}
   s.email             = %q{info@refinerycms.com}
   s.authors           = ["Philip Arndt", "Uģis Ozols", "Joe Sak"]
+  s.metadata          = { "contributor" => "Brice Sanchez <https://github.com/bricesanchez>" }
```
Comment (parndt) [inline refinerycms-search.gemspec]: Just add yourself as an author :smile:

**Comment by parndt** [comment]
@bricesanchez reviewed, thanks!

File: Rakefile [hunk 10]
```diff
@@ -17,4 +17,4 @@ Refinery::Testing::Railtie.load_dummy_tasks(ENGINE_PATH)
 
 load File.expand_path('../tasks/rspec.rake', __FILE__)
 
-task :default => :spec
+task default: :spec
```
Comment (simi) [inline Rakefile]: :question:

File: app/views/refinery/search/search/show.html.erb (hunk 2 above)
Comment (simi) [inline app/views/refinery/search/search/show.html.erb]: Yup, always be positive.

File: db/seeds.rb [hunk 11]
```diff
@@ -1,23 +1,19 @@
...
-if defined?(Refinery::Page)
-  unless Refinery::Page.where(:menu_match => "^/search.*$").any?
-    page = Refinery::Page.create(
-      :title => "Search Results",
-      :show_in_menu => false,
-      :link_url => "/search",
-      :deletable => false,
-      :menu_match => "^/search.*$"
-    )
+if defined?(Refinery::Page) and !Refinery::Page.exists?(link_url: (url = Refinery::Search.page_url))
```
Comment (simi) [inline db/seeds.rb]: What's the usecase with missing `Refinery::Page`?

File: Gemfile [hunk 12]
```diff
@@ -10,6 +10,12 @@ git 'https://github.com/refinery/refinerycms', :branch => 'master' do
   end
 end
 
+group :test do
```
Comment (simi) [inline Gemfile]: Why was this moved?

Comment (bricesanchez) [inline Gemfile]: It was moved to fit the order of the Gemfile of refinerycms-blog

File: Rakefile (hunk 10 above)
Comment (bricesanchez) [inline Rakefile]: new hash syntax ?

File: Gemfile (hunk 12 above)
Comment (simi) [inline Gemfile]: hmm, that's an funny argument :smiley_cat:

File: Rakefile (hunk 10 above)
Comment (simi) [inline Rakefile]: hmm, I was thinking just about to use new hash syntax for new code

**Comment by simi** [comment]
What about backwards compatibility?

File: app/controllers/refinery/search/search_controller.rb (hunk 1 above)
Comment (bricesanchez) [inline app/controllers/refinery/search/search_controller.rb]: Ok, i didn't know that before_action is the new syntax : https://github.com/rails/rails/commit/9d62e04838f01f5589fa50b0baa480d60c815e2c. Thanks :)

File: app/views/refinery/search/search/show.html.erb (hunk 2 above)
Comment (bricesanchez) [inline app/views/refinery/search/search/show.html.erb]: Be happy ! Don't worry ! :microphone: :notes:

File: db/seeds.rb (hunk 11 above)
Comment (bricesanchez) [inline db/seeds.rb]: There is no usecase at this time, 

If `Refinery::Page` is missing, the search page will be not created.

It's like the `refinerycms-blog` engine.

File: refinerycms-search.gemspec (hunk 9 above)
Comment (bricesanchez) [inline refinerycms-search.gemspec]: No, it's because you've done a great job in the CMS that i'm able to improve this extension, i'm just a contributor ;)

Comment (simi) [inline refinerycms-search.gemspec]: So please remove this line. I don't want to keep contributors list in gemspec.

We have list of contributors available at https://github.com/refinery/refinerycms-search/graphs/contributors.

File: config/routes.rb (hunk 3 above)
Comment (bricesanchez) [inline config/routes.rb]: Yes, i changed it.

**Comment by bricesanchez** [comment]
@parndt Yeah, i'm not really proud of the number of commits... sorry.
@simi I don't know what to think about deprecations, most of features are new features. The main change who could break by update is the use of namespaces.

File: refinerycms-search.gemspec (hunk 9 above)
Comment (bricesanchez) [inline refinerycms-search.gemspec]: Done :)

**Comment by bricesanchez** [comment]
output_ebnhpj

**Comment by simi** [comment]
@bricesanchez the question is how to upgrade from older version to this on project. Do we need to notice something somewhere? Was public API changed?

**Comment by bricesanchez** [comment]
@simi, The only method i've deleted is : https://github.com/refinery/refinerycms-search/blob/2-1-stable/lib/refinery/search/engine.rb#L21

How to deprecate this method if we now use a config in an initializer ?

**Comment by simi** [comment]
Stub it and recreate it is functionality to use current API.

File: db/seeds.rb (hunk 11 above)
Comment (simi) [inline db/seeds.rb]: The question is how the `Refinery::Page` can be missing?

**Comment by bricesanchez** [comment]
Done :)

File: Rakefile (hunk 10 above)
Comment (parndt) [inline Rakefile]: preferably we don't change existing code unless we're already changing that line.
